Source/Core/VideoCommon/Assets/CustomResourceManager.h

```diff
@@ -0,0 +1,182 @@
+      Unloaded
+    };
+    LoadType load_type = LoadType::PendingReload;
+    bool has_errors = false;
+  };
+
+  struct InternalTextureDataResource
+  {
+    AssetData* asset_data = nullptr;
+    VideoCommon::GameTextureAsset* asset = nullptr;
+    std::shared_ptr<TextureData> texture_data;
+  };
+
+  void LoadTextureDataAsset(const CustomAssetLibrary::AssetID& asset_id,
+                            std::shared_ptr<VideoCommon::CustomAssetLibrary> library,
+                            InternalTextureDataResource* internal_texture_data);
+
+  template <typename T>
+  T* CreateAsset(const CustomAssetLibrary::AssetID& asset_id, AssetData::AssetType asset_type,
+                 std::shared_ptr<VideoCommon::CustomAssetLibrary> library)
+  {
+    const auto [it, added] =
+        m_asset_id_to_session_id.try_emplace(asset_id, m_session_id_to_asset_data.size());
+
+    if (added)
+    {
+      AssetData asset_data;
+      asset_data.asset = std::make_unique<T>(library, asset_id, it->second);
+      asset_data.type = asset_type;
+      asset_data.has_errors = false;
+      asset_data.load_type = AssetData::LoadType::PendingReload;
+      asset_data.load_request_time = {};
+
+      m_session_id_to_asset_data.insert_or_assign(it->second, std::move(asset_data));
+
+      // Synchronize the priority cache session id
+      m_pending_assets.prepare();
+      m_loaded_assets.prepare();
+    }
+    auto& asset_data_from_session = m_session_id_to_asset_data[it->second];
+
+    // Asset got unloaded, rebuild it with the same metadata
+    if (!asset_data_from_session.asset)
+    {
+      asset_data_from_session.asset = std::make_unique<T>(library, asset_id, it->second);
+      asset_data_from_session.has_errors = false;
+      asset_data_from_session.load_type = AssetData::LoadType::PendingReload;
+    }
+
+    return static_cast<T*>(asset_data_from_session.asset.get());
+  }
+
+  class LeastRecentlyUsedCache
+  {
+  public:
+    const std::list<CustomAsset*>& elements() const { return m_asset_cache; }
+
+    void put(u64 asset_session_id, CustomAsset* asset)
+    {
+      erase(asset_session_id);
+      m_asset_cache.push_front(asset);
+      m_iterator_lookup[m_asset_cache.front()->GetSessionId()] = m_asset_cache.begin();
+    }
+
+    CustomAsset* pop()
+    {
+      if (m_asset_cache.empty()) [[unlikely]]
+        return nullptr;
+      const auto ret = m_asset_cache.back();
+      if (ret != nullptr)
+      {
+        m_iterator_lookup[ret->GetSessionId()].reset();
+      }
+      m_asset_cache.pop_back();
+      return ret;
+    }
+
+    void prepare() { m_iterator_lookup.push_back(std::nullopt); }
+
+    void erase(u64 asset_session_id)
+    {
+      if (const auto iter = m_iterator_lookup[asset_session_id])
+      {
+        m_asset_cache.erase(*iter);
+        m_iterator_lookup[asset_session_id].reset();
+      }
+    }
+
+    bool empty() const { return m_asset_cache.empty(); }
+
+    std::size_t size() const { return m_asset_cache.size(); }
```

Comment:
Is this trying to be a stdlib compatible interface? Otherwise, I think the function names should be `PascalCase`.